fdbserver/ClusterController.actor.cpp

```diff
@@ -3295,108 +3365,166 @@ void checkOutstandingStorageRequests(ClusterControllerData* self) {
 	}
 }
 
-void checkBetterDDOrRK(ClusterControllerData* self) {
+// Finds and returns a new process for role
+WorkerDetails findNewProcessForSingleton(ClusterControllerData* self,
+                                         const ProcessClass::ClusterRole role,
+                                         std::map<Optional<Standalone<StringRef>>, int>& id_used) {
+	// find new process in cluster for role
+	WorkerDetails newWorker =
+	    self->getWorkerForRoleInDatacenter(
+	            self->clusterControllerDcId, role, ProcessClass::NeverAssign, self->db.config, id_used, {}, true)
+	        .worker;
+
+	// check if master's process is actually better suited for role
+	if (self->onMasterIsBetter(newWorker, role)) {
+		newWorker = self->id_worker[self->masterProcessId.get()].details;
+	}
+
+	// acknowledge that the pid is now potentially used by this role as well
+	id_used[newWorker.interf.locality.processId()]++;
+
+	return newWorker;
+}
+
+// Return best possible fitness for singleton. Note that lower fitness is better.
+ProcessClass::Fitness findBestFitnessForSingleton(const ClusterControllerData* self,
+                                                  const WorkerDetails& worker,
+                                                  const ProcessClass::ClusterRole& role) {
+	auto bestFitness = worker.processClass.machineClassFitness(role);
+	// If the process has been marked as excluded, we take the max with ExcludeFit to ensure its fit
+	// is at least as bad as ExcludeFit. This assists with successfully offboarding such processes
+	// and removing them from the cluster.
+	if (self->db.config.isExcludedServer(worker.interf.addresses())) {
+		bestFitness = std::max(bestFitness, ProcessClass::ExcludeFit);
+	}
+	return bestFitness;
+}
+
+// Returns true iff the singleton is healthy. "Healthy" here means that
+// the singleton is stable (see below) and doesn't need to be rerecruited.
+// Side effects: (possibly) initiates recruitment
+template <class Interface>
+bool isHealthySingleton(ClusterControllerData* self,
```

Comment:
Maybe rename to tryRerecruitSingleton, so that the bool return val makes a little more sense?
It's kinda weird that it returns whether or not the singleton is healthy, but the logic for healthy and re-recruitment go hand-in-hand.

Update: Renamed to isHealthySingleton to be more clear, and since recruitment is listed as a side effect. Could still be improved.